src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -894,103 +925,337 @@ private bool CheckPropertyValueKind(SyntaxNode node, BoundExpression expr, BindV
-                        indexerAccess.ArgsToParamsOpt,
-                        checkingReceiver,
-                        diagnostics);
+                        if (parameters[paramIndex].RefKind == RefKind.RefReadOnly)
+                        {
+                            effectiveRefKind = RefKind.RefReadOnly;
+                            inParametersMatchedWithArgs = inParametersMatchedWithArgs ?? ArrayBuilder<bool>.GetInstance(parameters.Length, fillWithValue: false);
+                            inParametersMatchedWithArgs[paramIndex] = true;
+                        }
+                    }
+
+                    // if ref is passed and it is not a ref-like, check ref escape, 
+                    // since it is the same or narrower than val escape and callee may wrap the ref into a val
+                    var argument = args[argIndex];
+                    var argEscape = effectiveRefKind != RefKind.None && argument.Type.IsByRefLikeType == false ?
+                                        GetRefEscape(args[argIndex], scopeOfTheContainingExpression) :
+                                        GetValEscape(args[argIndex], scopeOfTheContainingExpression);
+
+                    escapeScope = Math.Max(escapeScope, argEscape);
+
+                    if (escapeScope >= scopeOfTheContainingExpression)
+                    {
+                        // can't get any worse
+                        inParametersMatchedWithArgs?.Free();
+                        return escapeScope;
+                    }
                 }
             }
 
-            return true;
+            // handle omitted optional "in" parameters if there are any
+            if (!parameters.IsDefault)
+            {
+                for (int i = 0; i < parameters.Length; i++)
+                {
+                    var parameter = parameters[i];
+                    if (parameter.IsParams)
+                    {
+                        break;
+                    }
+
+                    // unmatched "in" parameter is same as a literal, its ref escape is scopeOfTheContainingExpression  (can't get any worse)
+                    //                                                its val escape is ExternalScope                   (does not affect overal result)
+                    if (parameter.RefKind == RefKind.RefReadOnly && 
+                        inParametersMatchedWithArgs?[i] != true &&
+                        parameter.Type.IsByRefLikeType == false)
+                    {
+                        inParametersMatchedWithArgs?.Free();
+                        return scopeOfTheContainingExpression;
+                    }
+                }
+            }
+
+            inParametersMatchedWithArgs?.Free();
+
+            // check receiver if ref-like
+            if (receiverOpt?.Type?.IsByRefLikeType == true)
+            {
+                return GetValEscape(receiverOpt, scopeOfTheContainingExpression);
+            }
+
+            return escapeScope;
         }
```

Comment:
Consider extracting a helper for checking `parameter` in the second loop as well.